I tried to test this on Linux -next (which is enabling -Wthread-safety): https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/

But when applying this patch I get tons of false positives:
```
  CC      lib/test_context-analysis.o
lib/test_context-analysis.c:84:1: error: releasing raw_spinlock '__UNIQUE_ID_unlock_245' that was not held [-Werror,-Wthread-safety-analysis]
   84 | TEST_SPINLOCK_COMMON(raw_spinlock,
      | ^
lib/test_context-analysis.c:76:5: note: expanded from macro 'TEST_SPINLOCK_COMMON'
   76 |                 { guard(class)(&d->lock);               op(d->counter); }                       \
      |                   ^
./include/linux/cleanup.h:419:2: note: expanded from macro 'guard'
  419 |         CLASS(_name, __UNIQUE_ID(guard))
      |         ^
./include/linux/cleanup.h:300:3: note: expanded from macro 'CLASS'
  300 |                 class_##_name##_constructor
      |                 ^
note: (skipping 4 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
././include/linux/compiler_types.h:16:23: note: expanded from macro '__PASTE'
   16 | #define __PASTE(a, b) ___PASTE(a, b)
      |                       ^
././include/linux/compiler_types.h:15:24: note: expanded from macro '___PASTE'
   15 | #define ___PASTE(a, b) a##b
      |                        ^
<scratch space>:29:1: note: expanded from here
   29 | __UNIQUE_ID_unlock_245
      | ^
lib/test_context-analysis.c:84:1: error: acquiring raw_spinlock 'd->lock' that is already held [-Werror,-Wthread-safety-analysis]
lib/test_context-analysis.c:77:5: note: expanded from macro 'TEST_SPINLOCK_COMMON'
   77 |                 { guard(class##_irq)(&d->lock);         op(d->counter); }                       \
      |                   ^
./include/linux/cleanup.h:419:2: note: expanded from macro 'guard'
  419 |         CLASS(_name, __UNIQUE_ID(guard))
      |         ^
./include/linux/cleanup.h:300:3: note: expanded from macro 'CLASS'
  300 |                 class_##_name##_constructor
      |                 ^
```

I haven't been able to isolate a minimal reproducer yet that we can use as a test case.

--- a/clang/lib/Analysis/ThreadSafety.cpp
+++ b/clang/lib/Analysis/ThreadSafety.cpp
@@ -2277,7 +2277,6 @@ void BuildLockset::VisitUnaryOperator(const UnaryOperator *UO) {
 void BuildLockset::VisitBinaryOperator(const BinaryOperator *BO) {
   if (!BO->isAssignmentOp())
     return;
-  
   checkAccess(BO->getLHS(), AK_Written);
   updateLocalVarMapCtx(BO);
 }

--- a/clang/lib/Analysis/ThreadSafety.cpp
+++ b/clang/lib/Analysis/ThreadSafety.cpp
@@ -1727,20 +1727,25 @@ class BuildLockset : public ConstStmtVisitor<BuildLockset> {
   FactSet FSet;
   // The fact set for the function on exit.
   const FactSet &FunctionExitFSet;
+  // Use `LVarCtx` to keep track of the context at each program point, which
+  // will be used to translate DREs (by `SExprBuilder::translateDeclRefExpr`)
+  // to their canonical definitions:
   LocalVariableMap::Context LVarCtx;
   unsigned CtxIndex;
 
   // To update and adjust the context.
   void updateLocalVarMapCtx(const Stmt *S) {
+    if (Analyzer->Handler.issueBetaWarnings()) {
+      // The lookup closure needs to be reconstructed with the LVarCtx at the
+      // point right before 'S'.  This is the context used for visiting 'S'.
+      Analyzer->SxBuilder.setLookupLocalVarExpr(
+          [this, Ctx = LVarCtx](const NamedDecl *D) mutable -> const Expr * {
+            return Analyzer->LocalVarMap.lookupExpr(D, Ctx);
+          });
+    }
     if (S)
+      // Update the context to the point right after 'S'.
       LVarCtx = Analyzer->LocalVarMap.getNextContext(CtxIndex, S, LVarCtx);
-    if (!Analyzer->Handler.issueBetaWarnings())
-      return;
-    // The lookup closure needs to be reconstructed with the refreshed LVarCtx.
-    Analyzer->SxBuilder.setLookupLocalVarExpr(
-        [this, Ctx = LVarCtx](const NamedDecl *D) mutable -> const Expr * {
-          return Analyzer->LocalVarMap.lookupExpr(D, Ctx);
-        });
   }
 
   // helper functions
@@ -2842,6 +2847,7 @@ void ThreadSafetyAnalyzer::runAnalysis(AnalysisDeclContext &AC) {
 
         case CFGElement::CleanupFunction: {
           const CFGCleanupFunction &CF = BI.castAs<CFGCleanupFunction>();
+          LocksetBuilder.updateLocalVarMapCtx(nullptr);
           LocksetBuilder.handleCall(
               /*Exp=*/nullptr, CF.getFunctionDecl(),
               SxBuilder.translateVariable(CF.getVarDecl(), nullptr),

--- a/clang/lib/Analysis/ThreadSafety.cpp
+++ b/clang/lib/Analysis/ThreadSafety.cpp
@@ -1728,24 +1728,23 @@ class BuildLockset : public ConstStmtVisitor<BuildLockset> {
   // The fact set for the function on exit.
   const FactSet &FunctionExitFSet;
   // Use `LVarCtx` to keep track of the context at each program point, which
-  // will be used to translate DREs (by `SExprBuilder::translateDeclRefExpr`)
+  // will be used to translate variables (by `SExprBuilder::translateVariable`)
   // to their canonical definitions:
   LocalVariableMap::Context LVarCtx;
   unsigned CtxIndex;
 
-  // To update and adjust the context.
+  // To update the context used in attr-expr translation.  If `S` is non-null,
+  // the context is updated to the program point right after 'S'.
   void updateLocalVarMapCtx(const Stmt *S) {
-    if (Analyzer->Handler.issueBetaWarnings()) {
-      // The lookup closure needs to be reconstructed with the LVarCtx at the
-      // point right before 'S'.  This is the context used for visiting 'S'.
-      Analyzer->SxBuilder.setLookupLocalVarExpr(
-          [this, Ctx = LVarCtx](const NamedDecl *D) mutable -> const Expr * {
-            return Analyzer->LocalVarMap.lookupExpr(D, Ctx);
-          });
-    }
     if (S)
-      // Update the context to the point right after 'S'.
       LVarCtx = Analyzer->LocalVarMap.getNextContext(CtxIndex, S, LVarCtx);
+    if (!Analyzer->Handler.issueBetaWarnings())
+      return;
+    // The lookup closure needs to be reconstructed with the refreshed LVarCtx.
+    Analyzer->SxBuilder.setLookupLocalVarExpr(
+        [this, Ctx = LVarCtx](const NamedDecl *D) mutable -> const Expr * {
+          return Analyzer->LocalVarMap.lookupExpr(D, Ctx);
+        });
   }
 
   // helper functions
@@ -2278,9 +2277,9 @@ void BuildLockset::VisitUnaryOperator(const UnaryOperator *UO) {
 void BuildLockset::VisitBinaryOperator(const BinaryOperator *BO) {
   if (!BO->isAssignmentOp())
     return;
-
-  updateLocalVarMapCtx(BO);
+  
   checkAccess(BO->getLHS(), AK_Written);
+  updateLocalVarMapCtx(BO);
 }
 
 /// Whenever we do an LValue to Rvalue cast, we are reading a variable and
@@ -2325,8 +2324,6 @@ void BuildLockset::examineArguments(const FunctionDecl *FD,
 }
 
 void BuildLockset::VisitCallExpr(const CallExpr *Exp) {
-  updateLocalVarMapCtx(Exp);
-
   if (const auto *CE = dyn_cast<CXXMemberCallExpr>(Exp)) {
     const auto *ME = dyn_cast<MemberExpr>(CE->getCallee());
     // ME can be null when calling a method pointer
@@ -2390,9 +2387,9 @@ void BuildLockset::VisitCallExpr(const CallExpr *Exp) {
   }
 
   auto *D = dyn_cast_or_null<NamedDecl>(Exp->getCalleeDecl());
-  if (!D)
-    return;
-  handleCall(Exp, D);
+  if (D)
+    handleCall(Exp, D);
+  updateLocalVarMapCtx(Exp);
 }
 
 void BuildLockset::VisitCXXConstructExpr(const CXXConstructExpr *Exp) {
@@ -2421,8 +2418,6 @@ static const Expr *UnpackConstruction(const Expr *E) {
 }
 
 void BuildLockset::VisitDeclStmt(const DeclStmt *S) {
-  updateLocalVarMapCtx(S);
-
   for (auto *D : S->getDeclGroup()) {
     if (auto *VD = dyn_cast_or_null<VarDecl>(D)) {
       const Expr *E = VD->getInit();
@@ -2442,6 +2437,7 @@ void BuildLockset::VisitDeclStmt(const DeclStmt *S) {
       }
     }
   }
+  updateLocalVarMapCtx(S);
 }
 
 void BuildLockset::VisitMaterializeTemporaryExpr(
@@ -2847,7 +2843,6 @@ void ThreadSafetyAnalyzer::runAnalysis(AnalysisDeclContext &AC) {
 
         case CFGElement::CleanupFunction: {
           const CFGCleanupFunction &CF = BI.castAs<CFGCleanupFunction>();
-          LocksetBuilder.updateLocalVarMapCtx(nullptr);
           LocksetBuilder.handleCall(
               /*Exp=*/nullptr, CF.getFunctionDecl(),
               SxBuilder.translateVariable(CF.getVarDecl(), nullptr),

--- a/clang/test/SemaCXX/warn-thread-safety-analysis.cpp
+++ b/clang/test/SemaCXX/warn-thread-safety-analysis.cpp
@@ -7493,7 +7493,21 @@ void testPointerAliasEscapeMultiple(Foo *F) {
     escapeAliasMultiple(&L, &L, &Fp);
     Fp->mu.Unlock(); // expected-warning{{releasing mutex 'Fp->mu' that was not held}}
 } // expected-warning{{mutex 'F->mu' is still held at the end of function}}
-  
+
+void unlockFooWithEscapablePointer(Foo **Fp) EXCLUSIVE_UNLOCK_FUNCTION((*Fp)->mu);
+void testEscapeInvalidationHappensRightAfterTheCall(Foo* F) {
+    Foo* L;
+    L = F;
+    L->mu.Lock();
+    // Release the lock hold by 'L' before clear its definition.
+    unlockFooWithEscapablePointer(&L);
+}
+
+void testCleanUpFunctionWithLocalVarUpdated(Foo* F) {
+  F->mu.Lock();
+  Foo * __attribute__((unused, cleanup(unlockFooWithEscapablePointer))) L = F;
+}
+
 void testPointerAliasTryLock1() {
   Foo *ptr = returnsFoo();
   if (ptr->mu.TryLock()) {

--- a/clang/test/SemaCXX/warn-thread-safety-analysis.cpp
+++ b/clang/test/SemaCXX/warn-thread-safety-analysis.cpp
@@ -7503,6 +7503,18 @@ void testEscapeInvalidationHappensRightAfterTheCall(Foo* F) {
     unlockFooWithEscapablePointer(&L);
 }
 
+
+void testEscapeInvalidationHappensRightAfterTheCtorCall(Foo* F) {
+  Foo* L = F;
+  MutexLock ScopeLock(&L->mu);
+
+  struct {
+    int DataMember GUARDED_BY(F->mu);
+  } Data;
+
+  Data.DataMember = 0;
+}
+
 void testCleanUpFunctionWithLocalVarUpdated(Foo* F) {
   F->mu.Lock();
   Foo * __attribute__((unused, cleanup(unlockFooWithEscapablePointer))) L = F;

